File: python/src/epostak/resources/extract.py
"""Extract resource -- AI-powered OCR extraction from PDFs and images.

Provides :class:`ExtractResource` for extracting structured invoice data from
uploaded files.  Requires the Enterprise plan.  Supported formats:
``application/pdf``, ``image/jpeg``, ``image/png``, ``image/webp`` (max 20 MB).
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Sequence, Tuple, Union, TYPE_CHECKING

if TYPE_CHECKING:
    from epostak.types import BatchExtractResult, ExtractResult

from epostak.resources.documents import _BaseResource


class ExtractResource(_BaseResource):
    """AI-powered OCR extraction from PDFs and images.

    Requires Enterprise plan. Supported MIME types: ``application/pdf``,
    ``image/jpeg``, ``image/png``, ``image/webp``. Max 20 MB per file.
    """

    def single(
        self,
        file: bytes,
        mime_type: str,
        file_name: str = "document",
    ) -> ExtractResult:
        """Extract structured data from a single file.

        Args:
            file: File content as raw bytes.
            mime_type: MIME type -- ``"application/pdf"``, ``"image/jpeg"``,
                ``"image/png"``, or ``"image/webp"``.
            file_name: Display name for the file (default ``"document"``).

        Returns:
            Dict with ``extraction`` (structured data), ``ubl_xml`` (generated UBL),
            ``confidence`` (0.0--1.0), and ``file_name``.

        Example::

            with open("invoice.pdf", "rb") as f:
                pdf_bytes = f.read()
            result = client.extract.single(pdf_bytes, "application/pdf", "invoice.pdf")
            print(result["confidence"], result["extraction"])
        """
        files = [("file", (file_name, file, mime_type))]
        return self._request("POST", "/extract", files=files)

    def batch(
        self,
        files: List[Dict[str, Any]],
    ) -> BatchExtractResult:
        """Extract structured data from multiple files (up to 10).

        Args:
            files: List of dicts, each with keys ``file`` (bytes), ``mime_type`` (str),
                and optional ``file_name`` (str, defaults to ``"document"``).

        Returns:
            Dict with ``batch_id``, ``total``, ``successful``, ``failed``, and ``results``
            (one entry per file with ``extraction``, ``ubl_xml``, ``confidence``, or ``error``).

        Example::

            result = client.extract.batch([
                {"file": pdf_bytes, "mime_type": "application/pdf", "file_name": "inv1.pdf"},
                {"file": img_bytes, "mime_type": "image/png", "file_name": "inv2.png"},
            ])
            print(f"{result['successful']}/{result['total']} successful")
        """
        upload_files: List[Tuple[str, Tuple[str, bytes, str]]] = []
        for f in files:
            name = f.get("file_name", "document")
            upload_files.append(("files", (name, f["file"], f["mime_type"])))
        return self._request("POST", "/extract/batch", files=upload_files)

File: ruby/lib/epostak/resources/extract.rb
# frozen_string_literal: true

require "stringio"

module EPostak
  module Resources
    # Resource for AI-powered data extraction from PDF invoices and scanned images.
    # Extracts structured invoice data (parties, line items, totals) and generates
    # UBL 2.1 XML that can be sent via Peppol.
    #
    # @example
    #   result = client.extract.single(File.open("invoice.pdf", "rb"), "application/pdf", file_name: "invoice.pdf")
    #   puts result["confidence"] # => 0.95
    #   puts result["ubl_xml"]   # => Generated UBL XML
    class Extract
      # @param http [EPostak::HttpClient] Internal HTTP client
      def initialize(http)
        @http = http
      end

      # Extract structured data from a single PDF or image file.
      # Uses AI-powered OCR to identify invoice fields, line items, parties,
      # and totals, then generates UBL 2.1 XML from the extracted data.
      #
      # @param file_path_or_io [String, IO, StringIO] File path string, or an IO-like object with read support
      # @param mime_type [String] MIME type of the file (e.g. "application/pdf", "image/png")
      # @param file_name [String] Original file name for reference (default: "document")
      # @return [Hash] Extracted data with "confidence" score and "ubl_xml" string
      #
      # @example From a file path
      #   result = client.extract.single("invoice.pdf", "application/pdf")
      #   if result["confidence"] > 0.8
      #     client.documents.send_document(receiverPeppolId: "0245:1234567890", xml: result["ubl_xml"])
      #   end
      #
      # @example From an IO object
      #   io = File.open("scan.png", "rb")
      #   result = client.extract.single(io, "image/png", file_name: "scan.png")
      def single(file_path_or_io, mime_type, file_name: "document")
        io = resolve_io(file_path_or_io)
        @http.request_multipart("/extract", [
          { field: "file", io: io, mime_type: mime_type, filename: file_name }
        ])
      end

      # Extract structured data from multiple PDF or image files in a single request.
      # Each file is processed independently -- individual failures don't affect others.
      #
      # @param files [Array<Hash>] Array of file descriptors, each with:
      #   - +:file+ [String, IO] - File path or IO object
      #   - +:mime_type+ [String] - MIME type
      #   - +:file_name+ [String] - Optional file name (default: "document")
      # @return [Hash] Batch result with "results" array, "successful" and "total" counts
      #
      # @example
      #   result = client.extract.batch([
      #     { file: "invoice1.pdf", mime_type: "application/pdf", file_name: "invoice1.pdf" },
      #     { file: "invoice2.pdf", mime_type: "application/pdf", file_name: "invoice2.pdf" }
      #   ])
      #   puts "#{result['successful']}/#{result['total']} extracted successfully"
      def batch(files)
        parts = files.map do |f|
          {
            field: "files",
            io: resolve_io(f[:file]),
            mime_type: f[:mime_type],
            filename: f[:file_name] || "document"
          }
        end
        @http.request_multipart("/extract/batch", parts)
      end

      private

      # Convert a file path string to an IO object, or return an IO as-is.
      #
      # @param file_path_or_io [String, IO]
      # @return [IO]
      def resolve_io(file_path_or_io)
        if file_path_or_io.is_a?(String)
          File.open(file_path_or_io, "rb")
        else
          file_path_or_io
        end
      end
    end
  end
end
